### PR TITLE
361 transform thread unit tests into integration tests

### PR DESCRIPTION
**Pull request by AndreasHuber**

### CODE REVIEW

**Comment by hugotiburtino** [comment]
the error in pacts_tests doesn't have to do with your changes, we need to fix it in api: https://github.com/serlo/database-layer/actions/runs/4049109889/jobs/6965117034

**Review by hugotiburtino — CHANGES_REQUESTED** [review]
Thanks!
If we dry it a little more, we could merge it.

Side note: I would prefer to have different tests for each case. Flags and too much comments are code smells. After our conversation, I'm also convinced that tests are not place for logic.

Where do I see parametrized tests are most applicable: by testing macros or testing the almost same result with different things (v.g. UuidQuery for each kind of uuid).

File: server/tests/thread.rs [hunk 1]
```diff
@@ -267,4 +269,68 @@ mod thread_mutations {
...
     }
+
+    #[rstest]
+    // positive cases:
+    // no ID should be ok but do nothing
+    #[case(StatusCode::OK, false, [].as_slice(), 1, true)]
+    // single ID, should trigger an event for the change
+    #[case(StatusCode::OK, true, [17666].as_slice(), 1, true)]
+    // single ID, but no state change -> should not trigger event
+    #[case(StatusCode::OK, false, [17666].as_slice(), 1, false)]
```
[the remark this answers is not in the document]
Comment (AndreasHuber) [reply server/tests/thread.rs]: You were correct that I added things I wasn't using yet which one shouldn't, so now I also added negative test cases.

File: server/tests/thread.rs [hunk 2]
```diff
@@ -267,4 +269,68 @@ mod thread_mutations {
...
+                            "SetThreadStateNotificationEvent"
+                        );
+                        assert_eq!(latest_event["objectId"], *id);
+                        assert_eq!(latest_event["threadId"], *id);
+                        assert_eq!(latest_event["actorId"], user_id);
+                        assert_eq!(latest_event["archived"], archived);
+                        assert!(event_age < Duration::minutes(1));
+                    } else {
+                        assert!(event_age > Duration::minutes(1));
+                    }
```
[the remark this answers is not in the document]
Comment (AndreasHuber) [reply server/tests/thread.rs]: I think testing for event age is the most obvious way if you also consider the case where no event should be triggered. Alternatively one could look what is the last event currently and compare to its ID but I view this as more fragile or less obvious.

File: server/tests/thread.rs [hunk 3]
```diff
@@ -267,4 +269,68 @@ mod thread_mutations {
...
+
+            Message::new("EventsQuery", json!({ "first": 1 }))
+                .execute_on(&mut transaction)
+                .await
+                .should_be_ok_with(|result| {
+                    let latest_event = &result["events"][0];
+                    let event_age = DateTime::now().signed_duration_since(
+                        serde_json::from_value(latest_event["date"].clone()).unwrap(),
+                    );
+                    if should_trigger_event {
```
[the remark this answers is not in the document]
Comment (AndreasHuber) [reply server/tests/thread.rs]: We may also want to test that no event is triggered when none should be triggered, so we always need an EventsQuery.

**Comment by AndreasHuber** [comment]
Thanks for your input Hugo, it definitely made me think about how to best write test cases. 

I do agree that having if-statements in the test case is not optimal and I would prefer to not have such logic in a test case. 
The alternative to a parametrized test would be to copy paste the required parts for every case in a separate test. I think this would cause an amount of code that would be even harder to grasp or maintain. So if these are the two options, I tend to prefer the parametrized test still. 

With our current plans for the comments, we may change this again anyways, so maybe let's try to close it as soon as there is nothing objectively wrong and it presents an improvement over how it was before.

**Review by hugotiburtino — APPROVED** [review]
